firedrake/interpolation.py

```diff
@@ -961,12 +958,23 @@ def callable():
         if bcs and rank == 1:
             loops.extend(partial(bc.apply, f) for bc in bcs)
 
-        def callable(loops, f):
-            for l in loops:
-                l()
+        def callable(loops, f, access):
+            if access is op2.WRITE:
+                for l in loops:
+                    l()
+                return f
+            # We are repeatedly incrementing into the same Dat so intermediate halo exchanges
+            # can be skipped.
+            f.dat.local_to_global_begin(access)
+            with f.dat.frozen_halo(access):
```

Comment:
@connorjward please see the bug in the description. I grabbed this line from https://github.com/firedrakeproject/firedrake/blob/ec5492847ace646af6f83982a6e540ce32696a4a/firedrake/assemble.py#L1233

However that was not enough. To fix the reuse work in parallel also needed to wrap this block aroubnd `f.dat.local_to_global_[begin/end]`, however this breaks on the third call. Am I missing something?

Reply:
I am happy to try and change the order of the call to zero(). I can't find where that is done in assemble.py

Reply:
The fact that we're on a MixedFunctionSpace is not causing the issue with the reuse